pkg/kubeapiserver/authorizer/reload.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorizer
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	authzconfig "k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/authorization/authorizerfactory"
+	"k8s.io/apiserver/pkg/authorization/union"
+	"k8s.io/apiserver/pkg/server/options/authorizationconfig/metrics"
+	webhookutil "k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/apiserver/plugin/pkg/authorizer/webhook"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/auth/authorizer/abac"
+	"k8s.io/kubernetes/pkg/kubeapiserver/authorizer/modes"
+	"k8s.io/kubernetes/pkg/util/filesystem"
+	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/node"
+	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+)
+
+type reloadableAuthorizerResolver struct {
+	// initialConfig holds the ReloadFile used to initiate background reloading,
+	// and information used to construct webhooks that isn't exposed in the authorization
+	// configuration file (dial function, backoff settings, etc)
+	initialConfig Config
+
+	apiServerID string
+
+	reloadInterval         time.Duration
+	requireNonWebhookTypes sets.Set[authzconfig.AuthorizerType]
+
+	nodeAuthorizer *node.NodeAuthorizer
+	rbacAuthorizer *rbac.RBACAuthorizer
+	abacAuthorizer abac.PolicyList
+
+	lastLoadedLock   sync.Mutex
+	lastLoadedConfig *authzconfig.AuthorizationConfiguration
+	lastReadData     []byte
+
+	current atomic.Pointer[authorizerResolver]
+}
+
+type authorizerResolver struct {
+	authorizer   authorizer.Authorizer
+	ruleResolver authorizer.RuleResolver
+}
+
+func (r *reloadableAuthorizerResolver) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	return r.current.Load().authorizer.Authorize(ctx, a)
+}
+
+func (r *reloadableAuthorizerResolver) RulesFor(user user.Info, namespace string) ([]authorizer.ResourceRuleInfo, []authorizer.NonResourceRuleInfo, bool, error) {
+	return r.current.Load().ruleResolver.RulesFor(user, namespace)
+}
+
+// newForConfig constructs
+func (r *reloadableAuthorizerResolver) newForConfig(authzConfig *authzconfig.AuthorizationConfiguration) (authorizer.Authorizer, authorizer.RuleResolver, error) {
+	if len(authzConfig.Authorizers) == 0 {
```

Comment:
could there be an edge case where only a webhook authorizer was specified at startup and it got removed on reload?

Reply:
if they do that, it's invalid... you can't drop all authorizers